File: src/main/java/leetcode/hot100/LeetCode234.java
package leetcode.hot100;

import common.ListNode;

/**
 * @author aiwujingxin@gmail.com
 * @date 2022/9/11 23:43
 */
public class LeetCode234 {

    public static void main(String[] args) {
        ListNode node = new ListNode(1);
        ListNode node2 = new ListNode(2);
        ListNode node3 = new ListNode(3);
        ListNode node4 = new ListNode(4);
        node.next = node2;
        node2.next = node3;
        node3.next = node4;
        new LeetCode234().print_values_in_reverse(node);
    }

    private ListNode frontPointer;

    public boolean isPalindrome(ListNode head) {
        frontPointer = head;
        return recursivelyCheck(head);
    }

    private boolean recursivelyCheck(ListNode currentNode) {
        if (currentNode != null) {
            if (!recursivelyCheck(currentNode.next)) {
                return false;
            }
            if (currentNode.val != frontPointer.val) {
                return false;
            }
            frontPointer = frontPointer.next;
        }
        return true;
    }


    private void print_values_in_reverse(ListNode head) {
        if (head != null) {
            print_values_in_reverse(head.next);
            System.out.println(head.val);
        }
    }
}

File: src/main/GO/leetcode/lists/hot100/LeetCode234.go
package hot100

func isPalindrome(head *ListNode) bool {

	if head == nil || head.Next == nil {
		return true
	}
	normalNode := head
	var reverseCheck func(node *ListNode) bool
	reverseCheck = func(node *ListNode) bool {
		if node != nil {
			if !reverseCheck(node.Next) {
				return false
			}
			if normalNode.Val != node.Val { // 最开始是从这里比较的 比较的是头节点和尾节点
				return false
			}
			normalNode = normalNode.Next
		}
		return true
	}
	return reverseCheck(head)
}
